src/main/java/ratpack/pac4j/internal/Pac4jAuthenticator.java

```diff
@@ -110,11 +110,13 @@ public void handle(Context ctx) throws Exception {
     return Promise.value(new Clients(absoluteCallbackUrl, clients));
   }
 
-  private <C extends Credentials, U extends CommonProfile> Promise<U> getProfile(WebContext webContext, Client<C, U> client) throws HttpAction {
-    return Blocking.get(() -> {
-      C credentials = client.getCredentials(webContext);
-      return client.getUserProfile(credentials, webContext);
-    });
+  private Promise<Optional<UserProfile>> getProfile(WebContext webContext,
+      Client client) throws HttpAction {
+    return Blocking.get(
+        () -> client.getCredentials(webContext, ((RatpackWebContext) webContext).getSessionStore())
+            .flatMap(credentials -> client.getUserProfile(credentials, webContext,
+                ((RatpackWebContext) webContext).getSessionStore()))
+    );
```

Comment:
This might read a bit more cleanly if you store the ratpack-cast web context to a variable so you don't need to repeat the cast.

Reply:
Turns out the only is passing a ratpack context anyways, so I can get rid of the cast altogether.